### PR TITLE
Refactor VmapPhysicalView::newLogicalToPhysical

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49482 Refactor VmapPhysicalView::newLogicalToPhysical**

Motivation
==========
Batching rules always invoke newLogicalToPhysical at the very end to turn
a physical tensor into a logical BatchedTensor (an example is below):
```
Tensor select_backward_batching_rule(const Tensor& grad, IntArrayRef input_sizes, int64_t dim, int64_t index) {
  auto grad_physical = MultiBatchVmapTransform::logicalToPhysical(grad);
  auto grad_input = at::zeros(grad_physical.getPhysicalShape(input_sizes), grad.options());
  auto physical_dim = getGradInputPhysicalDim(dim, input_sizes, grad_physical.numBatchDims());
  grad_input.select(physical_dim, index).copy_(grad_physical.tensor());
  return grad_physical.newLogicalFromPhysical(grad_input);
}
```
However, @albanD noted that this function is confusing and ambiguous
because it's unclear which physical tensor is being turned into the logical
(in this case, grad_physical is a VmapPhysicalView, but we're really transforming
grad_input and returning it).
https://github.com/pytorch/pytorch/pull/44505#discussion_r487144018

I didn't want to make too many changes to the batching rule API because
I think we'll change it even more in the future, but this PR attempts to
remove the ambiguity by applying one of the suggestions in
https://github.com/pytorch/pytorch/pull/44505#discussion_r487144018

This PR
=======

The diagnosis of the problem is that we were conflating
"VmapPhysicalView", which maps logical attributes on a Tensor (like
dimension and shape) to physical attributes, with the reverse
physical-to-logical map. This PR creates a new VmapPhysicalToLogicalMap
object that handles the latter.

Instead of calling `grad_physical.newLogicalFromPhysical(grad_input)`,
an author of batching rules should now retrieve the VmapPhysicalToLogicalMap
object and apply it to their physical input. So the above code becomes:
```
grad_physical.getPhysicalToLogicalMap().apply(grad_input)
```

I've also moved VmapPhysicalView::makeLogicalFromPhysicalListInplace
to VmapPhysicalToLogicalMap::applyInplace.

Test Plan
=========
wait for tests

Differential Revision: [D25592645](https://our.internmc.facebook.com/intern/diff/D25592645)

### CODE REVIEW

**Comment by ezyang** [comment]
Going to let Alban review this one.

**Comment by albanD** [comment]
Hey,
This PR definitely helps for the confusion I had earlier.

Going back at your example:
```c++
Tensor select_backward_batching_rule(const Tensor& grad, IntArrayRef input_sizes, int64_t dim, int64_t index) {
  auto grad_physical = MultiBatchVmapTransform::logicalToPhysical(grad);
  auto grad_input = at::zeros(grad_physical.getPhysicalShape(input_sizes), grad.options());
  auto physical_dim = getGradInputPhysicalDim(dim, input_sizes, grad_physical.numBatchDims());
  grad_input.select(physical_dim, index).copy_(grad_physical.tensor());
  return grad_physical.newLogicalFromPhysical(grad_input);
}
```

I am still unsure why we need the physical Tensor and the struct to convert to be so tightly coupled. Could something like this be more readable:
```c++
Tensor select_backward_batching_rule(const Tensor& grad, IntArrayRef input_sizes, int64_t dim, int64_t index) {
  auto vmap_transform = MultiBatchVmapTransform::getFromTensor(grad);
  auto grad_physical = vmap_transform.newPhysicalFromLogical(grad); # This is an actual Tensor
  auto grad_input = at::zeros(vmap_transform.getPhysicalShape(input_sizes), grad.options());
  auto physical_dim = getGradInputPhysicalDim(dim, input_sizes, vmap_transform.numBatchDims());
  grad_input.select(physical_dim, index).copy_(grad_physical);
  return vmap_transform.newLogicalFromPhysical(grad_input);
}
```

Is the main reason you don't do this that the `vmap_transform` would need to store something like grad anyway?

**Comment by zou3519** [comment]
Thanks for the suggestion, @albanD! 

Here's my thinking: that example works when there is only one Tensor to transform. When there are multiple Tensors to transform, things get a bit hairy. Here's an example, but note that the batching rule of index_select hasn't been implemented yet so this is only a hypothetical example.

```
input = torch.randn(10, 2, 3, 4)
index = torch.tensor([1])
vmap(index_select, (0, None, None))(input, -1, index)
```

Here, hypothetically, we would like to apply a VmapTransform to both input and index:
```
Tensor index_select_batching_rule(const Tensor& input, int64_t dim, const Tensor& index) {
    auto vmap_transform = MultiBatchVmapTransform::get(input, index);
    auto physical_dim = vmap_transform.getPhysicalDim(dim);
    ...
}
```
Now, the question is, which tensor's physical dim are we actually talking about? If we're talking about `input`, the physical_dim is 3. If we're talking about `index`, the physical_dim is 0. To answer this question, we actually need to pass a Tensor to vmap_transform:
```
// Option 1
Tensor index_select_batching_rule(const Tensor& input, int64_t dim, const Tensor& index) {
    auto vmap_transform = MultiBatchVmapTransform::get(input, index);
    auto physical_input = vmap_transform.newPhysicalFromLogical(input);
    auto physical_index = vmap_transform.newPhysicalFromLogical(index);
    auto physical_dim = vmap_transform.getPhysicalDim(input, dim);
    ...
}
```
The same is true for `getPhysicalShape`.

One way around this, is to return a vmap_transform for every input:

```
// Option 2
Tensor index_select_batching_rule(const Tensor& input, int64_t dim, const Tensor& index) {
    auto [vmap_transform_input, vmap_transform_index] = MultiBatchVmapTransform::get(input, index);
    auto physical_input = vmap_transform_input.newPhysicalFromLogical(input);
    auto physical_index = vmap_transform_index.newPhysicalFromLogical(index);
    auto physical_dim = vmap_transform_input.getPhysicalDim(dim);
    ...
}
```
`vmap_transform_input`, `vmap_transform_index` are almost the same as the [VmapPhysicalView objects](https://github.com/pytorch/pytorch/blob/6db5e857269a5e68c59c7452c268bf4847d7aa29/aten/src/ATen/VmapTransforms.h#L97-L143), which hold some mapping information. The difference is that the VmapPhysicalView objects hold the actual physical tensors themselves.

Option 1, which is more similar to what you proposed, could be a reasonable solution. I think it would work out, and it does decouple the vmap transform from the actual physical tensors. However, because not many batching rules are implemented right now, I'm not sure if the approach will work for future batching rules (I also don't know if the current design will hold up against those). I'm inclined to go off and implement a few more batching rules and come back to the question. What do you think?

**Review by albanD — APPROVED** [review]
I was expecting issues for formulas with more than one Tensor.
I prefer option 2 here in the sense that it both provides one transform object per input and hides nicely the fact that each of them needs to know about the other.

But I agree with you that we most likely want to wait to see a few more formulas before locking into a given design.

This PR LGTM, it is a nice incremental improvemnt!
